File: c++/513. Find Bottom Left Tree Value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
        int res = 0;
        queue<TreeNode*> que;
        que.push(root);
        while (!que.empty()) {
            int n = que.size();
            res = que.front()->val;
            while (n-->0) {
                TreeNode*  p = que.front();
                que.pop();
                if (p->left)
                    que.push(p->left);
                if (p->right)
                    que.push(p->right);
            }
        }
        return res;
    }
};

File: go/src/leetcode/513/513. Find Bottom Left Tree Value.go
/**
@author ZhengHao Lou
Date    2022/06/22
*/
package main

/**
https://leetcode.cn/problems/find-bottom-left-tree-value/
*/
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findBottomLeftValue(root *TreeNode) (val int) {
	que := []*TreeNode{root}
	for len(que) != 0 {
		val = que[0].Val
		for l := len(que); l > 0; l-- {
			node := que[0]
			que = que[1:]
			if node.Left != nil {
				que = append(que, node.Left)
			}
			if node.Right != nil {
				que = append(que, node.Right)
			}
		}
	}
	return
}
